### PR TITLE
Add support for 1:1 job per topology assignment

### LINKED ISSUE #27: 1:1 job to topology assignment
Consider the case where in a replicatedJob each job needs to land on a separate topology, for example one per network domain, like a rack.

This can be achieved using a combination of affinity and anti-affinity constraints on the pod template:

```
      affinity:
        podAffinity:  # ensures the pods of this job land on the same rack
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: job-name
                operator: In
                values:
                - my-job
              topologyKey: rack
        podAntiAffinity: # ensures only this job lands on the rack
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: job-name
                operator: NotIn
                values:
                - my-job
              - key: job-name
                operator: Exists
            namespaceSelector: {}
            topologyKey: rack
```

The problem is that the affinity terms above require setting the value of the `job-name` label explicitly, which in a replicatedJob is different for each job replica.

[The MatchLabelKeys enhancement ](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3633-matchlabelkeys-to-podaffinity)in upstream k8s should address this problem, but it is not yet available.

Until this is available, we can have this injected by JobSet, triggered by an api we add to ReplicatedJob type that look like :

```
exclusive
  topologyKey: rack
```

### PR DESCRIPTION
**Pull request by danielvegamyhre**
Fixes #27 

Adds `Exclusive` struct to the API which has a `TopologyKey` field. If specified, the JobSet controller will append pod affinity and anti-affinity terms to each job pod template spec, which ensures 1:1 job per topology assignment, and ensures all pods for each job land on the same topology.

**Note**: I also had to update the `make install` command in the Makefile to use `kubectl create` instead of `kubectl apply` as a workaround for this [issue](https://github.com/kubernetes-sigs/kubebuilder/issues/1140).

Did a manual test on a cluster with 3 node pools, using the topology key `cloud.google.com/gke-nodepool` to place 1 job per node pool, and it worked. 

Test JobSet with 3 replicated jobs, 1 replica each:

```
apiVersion: batch.x-k8s.io/v1alpha1
kind: JobSet
metadata:
  name: jobpertopology
  labels:
    app.kubernetes.io/name: jobset
    app.kubernetes.io/instance: jobset-example
    app.kubernetes.io/part-of: project
    app.kubernetes.io/managed-by: kustomize
    app.kubernetes.io/created-by: project
spec:
  jobs:
  - name: worker0
    network: 
      enableDNSHostnames: true
    exclusive:
      topologyKey: cloud.google.com/gke-nodepool
    template:
      metadata:
        name: "worker1"
      spec:
        completionMode: Indexed
        parallelism: 2
        completions: 2
        template:
          spec:
            restartPolicy: Never
            containers:
            - name: busybox-short-sleep
              image: busybox:latest
              command:
              - sleep
              - "1000" 
  - name: worker1
    network: 
      enableDNSHostnames: true
    exclusive:
      topologyKey: cloud.google.com/gke-nodepool
    template:
      metadata:
        name: "worker1"
      spec:
        completionMode: Indexed
        parallelism: 2
        completions: 2
        template:
          spec:
            restartPolicy: Never
            containers:
            - name: busybox-short-sleep
              image: busybox:latest
              command:
              - sleep
              - "1000" 
  - name: worker2
    network: 
      enableDNSHostnames: true
    exclusive:
      topologyKey: cloud.google.com/gke-nodepool
    template:
      metadata:
        name: "worker2"
      spec:
        completionMode: Indexed
        parallelism: 2
        completions: 2
        template:
          spec:
            restartPolicy: Never
            containers:
            - name: busybox-short-sleep
              image: busybox:latest
              command:
              - sleep
              - "1000" 
```

Pod placement:
```
$ k get pods -o wide
NAME                               READY   STATUS    RESTARTS   AGE   IP         NODE                                     NOMINATED NODE   READINESS GATES
jobpertopology-worker0-0-0-6tg7d   1/1     Running   0          7s    10.8.7.3   gke-jobset4-np2-782cc200-vxq8            <none>           <none>
jobpertopology-worker0-0-1-262vf   1/1     Running   0          7s    10.8.8.3   gke-jobset4-np2-782cc200-946b            <none>           <none>
jobpertopology-worker1-0-0-lc2q6   1/1     Running   0          7s    10.8.5.4   gke-jobset4-np1-004ddd96-58v7            <none>           <none>
jobpertopology-worker1-0-1-4wgcz   1/1     Running   0          7s    10.8.3.4   gke-jobset4-np1-004ddd96-cvzx            <none>           <none>
jobpertopology-worker2-0-0-849c6   1/1     Running   0          7s    10.8.1.9   gke-jobset4-default-pool-238f9ca7-8s4k   <none>           <none>
jobpertopology-worker2-0-1-mdlqv   1/1     Running   0          7s    10.8.2.6   gke-jobset4-default-pool-238f9ca7-hc67   <none>           <none>
```

Test JobSet with 1 replicated job with 3 replicas:
```
apiVersion: batch.x-k8s.io/v1alpha1
kind: JobSet
metadata:
  name: jobpertopology
  labels:
    app.kubernetes.io/name: jobset
    app.kubernetes.io/instance: jobset-example
    app.kubernetes.io/part-of: project
    app.kubernetes.io/managed-by: kustomize
    app.kubernetes.io/created-by: project
spec:
  jobs:
  - name: worker
    network: 
      enableDNSHostnames: true
    exclusive:
      topologyKey: cloud.google.com/gke-nodepool
    replicas: 3
    template:
      metadata:
        name: "worker"
      spec:
        completionMode: Indexed
        parallelism: 2
        completions: 2
        template:
          spec:
            restartPolicy: Never
            containers:
            - name: busybox-short-sleep
              image: busybox:latest
              command:
              - sleep
              - "1000"
```

Pod placement:
```
$ k get pods -o wide
NAME                              READY   STATUS    RESTARTS   AGE   IP          NODE                                     NOMINATED NODE   READINESS GATES
jobpertopology-worker-0-0-tdbfp   1/1     Running   0          11s   10.8.8.4    gke-jobset4-np2-782cc200-946b            <none>           <none>
jobpertopology-worker-0-1-rzrnb   1/1     Running   0          11s   10.8.7.4    gke-jobset4-np2-782cc200-vxq8            <none>           <none>
jobpertopology-worker-1-0-pkqgt   1/1     Running   0          11s   10.8.4.4    gke-jobset4-np1-004ddd96-c0b6            <none>           <none>
jobpertopology-worker-1-1-fdgtn   1/1     Running   0          11s   10.8.3.5    gke-jobset4-np1-004ddd96-cvzx            <none>           <none>
jobpertopology-worker-2-0-q7wbr   1/1     Running   0          11s   10.8.1.10   gke-jobset4-default-pool-238f9ca7-8s4k   <none>           <none>
jobpertopology-worker-2-1-kp4n7   1/1     Running   0          11s   10.8.2.7    gke-jobset4-default-pool-238f9ca7-hc67   <none>           <none>
```

### CODE REVIEW

File: pkg/controllers/jobset_controller.go [hunk 1]
```diff
@@ -436,6 +442,47 @@ func (r *JobSetReconciler) failJobSet(ctx context.Context, js *jobset.JobSet) er
...
+
+	// Pod anti-affinity ensures exclusively this job lands on the topology, preventing multiple jobs per topology.
+	job.Spec.Template.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution = append(job.Spec.Template.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution,
+		corev1.PodAffinityTerm{
+			LabelSelector: &metav1.LabelSelector{MatchExpressions: []metav1.LabelSelectorRequirement{
+				{
+					Key:      "job-name",
+					Operator: metav1.LabelSelectorOpNotIn,
+					Values:   []string{job.Name, ""},
+				},
```
Comment (ahg-g) [inline pkg/controllers/jobset_controller.go:480]: we should add another match expression here to limit the elusiveness to be between jobs only:

```suggestion
			LabelSelector: &metav1.LabelSelector{MatchExpressions: []metav1.LabelSelectorRequirement{
				{
					Key:      "job-name",
					Operator: metav1.LabelSelectorOpNotIn,
					Values:   []string{job.Name, ""},
				},
				{
					Key:      "job-name",
					Operator: metav1.LabelSelectorOpExists,
				},
				
```

File: api/v1alpha1/jobset_types.go [hunk 2]
```diff
@@ -98,6 +101,11 @@ type Network struct {
 	EnableDNSHostnames *bool `json:"enableDNSHostnames,omitempty"`
 }
 
+type Exclusive struct {
+	// Topology key for pod affinity/anti-affinity.
+	TopologyKey string `json:"topologyKey,omitempty"`
```
Comment (ahg-g) [inline api/v1alpha1/jobset_types.go:106]: lets also add namespaceSelector

```
// A label query over the set of namespaces that exclusiveness applies to. Defaults to the job's namespace. 
// An empty selector ({}) matches all namespaces.
NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
```

File: pkg/controllers/jobset_controller_test.go [hunk 3]
```diff
@@ -101,3 +102,164 @@ func TestIsJobFinished(t *testing.T) {
...
+				t.Errorf("unexpected diff in pod affinity (-want/+got): %s", diff)
+			}
+			if diff := cmp.Diff(tc.wantAffinity.PodAntiAffinity, tc.job.Spec.Template.Spec.Affinity.PodAntiAffinity); diff != "" {
+				t.Errorf("unexpected diff in pod anti-affinity (-want/+got): %s", diff)
+			}
+		})
+	}
+}
+
+func testJob(jobName string) *batchv1.Job {
```
Comment (ahg-g) [inline pkg/controllers/jobset_controller_test.go:219]: we have MakeJob in wrappers

File: pkg/controllers/jobset_controller.go [hunk 4]
```diff
@@ -436,6 +442,47 @@ func (r *JobSetReconciler) failJobSet(ctx context.Context, js *jobset.JobSet) er
...
+			}},
+			TopologyKey: topologyKey,
```
Comment (ahg-g) [inline pkg/controllers/jobset_controller.go:482]:
```suggestion
			TopologyKey: topologyKey,
			NamespaceSelector: <namespace-selector>
```

File: pkg/controllers/jobset_controller.go [hunk 5]
```diff
@@ -436,6 +442,47 @@ func (r *JobSetReconciler) failJobSet(ctx context.Context, js *jobset.JobSet) er
...
+		})
+
+	// Pod anti-affinity ensures exclusively this job lands on the topology, preventing multiple jobs per topology.
+	job.Spec.Template.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution = append(job.Spec.Template.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution,
+		corev1.PodAffinityTerm{
+			LabelSelector: &metav1.LabelSelector{MatchExpressions: []metav1.LabelSelectorRequirement{
+				{
+					Key:      "job-name",
+					Operator: metav1.LabelSelectorOpNotIn,
+					Values:   []string{job.Name, ""},
```
Comment (ahg-g) [inline pkg/controllers/jobset_controller.go:479]:
```suggestion
					Values:   []string{job.Name},
```

File: pkg/controllers/jobset_controller_test.go (hunk 3 above)
Comment (danielvegamyhre) [reply pkg/controllers/jobset_controller_test.go:219]: I thought about that as well but the MakeJob wrapper is for a batchv1.JobTemplateSpec, not a batchv1.Job, and I didn't think it was worth making a wrapper for this unit test - or do you think we should just to keep the pattern consistent throughout the codebase?

Comment (danielvegamyhre) [reply pkg/controllers/jobset_controller_test.go]: I went ahead and updated the wrappers to have distinct `MakeJobTemplate` (for wrapping `batchv1.JobTemplateSpec`) and `MakeJob` for wrapping (`batchv1.Job)` and refactored the unit test to use the wrapper.

File: pkg/controllers/jobset_controller.go [hunk 6]
```diff
@@ -436,6 +441,53 @@ func (r *JobSetReconciler) failJobSet(ctx context.Context, js *jobset.JobSet) er
 	})
 }
 
+// Appends pod affinity/anti-affinity terms to the job pod template spec,
+// ensuring that exclusively one job runs per topology and that all pods
+// from each job land on the same topology.
```
Comment (ahg-g) [inline pkg/controllers/jobset_controller.go:446]:
```suggestion
// ensuring that exclusively one job runs per topology domain and that all pods
// from each job land on the same topology domain.
```

File: pkg/controllers/jobset_controller.go [hunk 7]
```diff
@@ -436,6 +441,53 @@ func (r *JobSetReconciler) failJobSet(ctx context.Context, js *jobset.JobSet) er
...
+		job.Spec.Template.Spec.Affinity = &corev1.Affinity{}
+	}
+	if job.Spec.Template.Spec.Affinity.PodAffinity == nil {
+		job.Spec.Template.Spec.Affinity.PodAffinity = &corev1.PodAffinity{}
+	}
+	if job.Spec.Template.Spec.Affinity.PodAntiAffinity == nil {
+		job.Spec.Template.Spec.Affinity.PodAntiAffinity = &corev1.PodAntiAffinity{}
+	}
+
+	// Pod affinity ensures the pods of this job land on the same topology.
```
Comment (ahg-g) [inline pkg/controllers/jobset_controller.go:458]:
```suggestion
	// Pod affinity ensures the pods of this job land on the same topology domain.
```

File: pkg/controllers/jobset_controller.go [hunk 8]
```diff
@@ -436,6 +441,53 @@ func (r *JobSetReconciler) failJobSet(ctx context.Context, js *jobset.JobSet) er
...
+					Key:      jobset.JobNameKey,
+					Operator: metav1.LabelSelectorOpIn,
+					Values:   []string{job.Name},
+				},
+			}},
+			TopologyKey:       topologyKey,
+			NamespaceSelector: nsSelector,
+		})
+
+	// Pod anti-affinity ensures exclusively this job lands on the topology, preventing multiple jobs per topology.
```
Comment (ahg-g) [inline pkg/controllers/jobset_controller.go:472]:
```suggestion
	// Pod anti-affinity ensures exclusively this job lands on the topology, preventing multiple jobs per topology domain.
```

File: pkg/controllers/jobset_controller_test.go [hunk 9]
```diff
@@ -101,3 +105,163 @@ func TestIsJobFinished(t *testing.T) {
...
+									Operator: metav1.LabelSelectorOpNotIn,
+									Values:   []string{"value-bar"},
+								},
+							}},
+							TopologyKey: "topology.kubernetes.io/zone",
+						},
+					},
+				},
+			}).Obj(),
+			nsSelector: &metav1.LabelSelector{},
```
Comment (ahg-g) [inline pkg/controllers/jobset_controller_test.go:193]: just to use something different
```suggestion
			nsSelector: nil
```

File: api/v1alpha1/jobset_types.go [hunk 10]
```diff
@@ -21,6 +21,7 @@ import (
 const (
 	JobIndexLabel string = "jobset.sigs.k8s.io/job-index"
 	RestartsLabel string = "jobset.sigs.k8s.io/restart-attempt"
+	JobNameKey    string = "job-name" // TODO: Migrate to the fully qualified label name.
```
Comment (ahg-g) [inline api/v1alpha1/jobset_types.go:24]:
```suggestion
	JobNameKey    string = "job-name" // TODO(#26): Migrate to the fully qualified label name.
```

**Review by ahg-g — COMMENTED** [review]
few nits

**Pull request merged**